src/gatt_connection.rs

```diff
@@ -0,0 +1,56 @@
+use crate::att_error::AttError;
+use core::fmt::Debug;
+
+use crate::bluetooth_address::BluetoothAddress;
+
+/// Represents a GATT connection from the peripheral perspective, specifically used as the API
+/// to allow the BLE peripheral device to respond to connected clients.
+pub trait GattConnection {
+  type SystemError: Debug;
+  type Responder: GattResponder + Debug;
```

Comment:
Hmm, I'm not 100% sure this will map well to all implementations, or even to the reference BlueR one.  Responder and Writer are both created and passed around with very different lifetimes than the handle.  It's even possible to have a race condition where you drop the handle but still have callbacks firing talking about Responder and Writer.  I'll need to think about this a little bit more, I think the core feedback is likely valid here that the type relationships aren't as concise as they could be.